File: sdk/javascript/src/services/worker.service.ts
/**
 * Worker 服务
 * 提供 Worker 节点管理
 */

import { HttpClient } from '../utils/http';
import { Worker, WorkerStatus } from '../types';

/**
 * Worker 服务类
 */
export class WorkerService {
  private readonly http: HttpClient;
  private readonly resource = 'workers';

  constructor(http: HttpClient) {
    this.http = http;
  }

  /**
   * 获取 Worker 列表
   */
  async list(): Promise<Worker[]> {
    const response = await this.http.get<{ workers: Worker[] }>(`/api/${this.resource}`);
    return response.workers || [];
  }

  /**
   * 获取 Worker 详情
   */
  async get(workerId: string): Promise<Worker> {
    return this.http.get<Worker>(`/api/${this.resource}/${workerId}`);
  }

  /**
   * 获取在线 Worker
   */
  async getOnline(): Promise<Worker[]> {
    const workers = await this.list();
    return workers.filter((w) => w.status === WorkerStatus.ONLINE);
  }

  /**
   * 获取可用 Worker（在线且负载未满）
   */
  async getAvailable(): Promise<Worker[]> {
    const workers = await this.list();
    return workers.filter(
      (w) => w.status === WorkerStatus.ONLINE && w.currentLoad < w.maxLoad
    );
  }

  /**
   * 获取指定项目的 Worker
   */
  async getByProject(projectKey: string): Promise<Worker[]> {
    const workers = await this.list();
    return workers.filter(
      (w) =>
        w.status === WorkerStatus.ONLINE &&
        (w.supportedProjects.includes(projectKey) || w.supportedProjects.includes('*'))
    );
  }

  /**
   * 获取 Worker 状态统计
   */
  async getStats(): Promise<{
    total: number;
    online: number;
    offline: number;
    busy: number;
  }> {
    const workers = await this.list();
    return {
      total: workers.length,
      online: workers.filter((w) => w.status === WorkerStatus.ONLINE).length,
      offline: workers.filter((w) => w.status === WorkerStatus.OFFLINE).length,
      busy: workers.filter((w) => w.status === WorkerStatus.BUSY).length,
    };
  }
}

File: sdk/python/src/clawkit/services/worker_service.py
"""
Worker 服务
"""

from typing import List
from clawkit.http import HttpClient
from clawkit.models.worker import Worker, WorkerStatus


class WorkerService:
    """Worker 服务类"""

    def __init__(self, http: HttpClient):
        self.http = http
        self.resource = "workers"

    def list(self) -> List[Worker]:
        """获取 Worker 列表"""
        data = self.http.get(f"/api/{self.resource}")
        workers = data.get("workers", []) if isinstance(data, dict) else data
        if isinstance(workers, list):
            return [Worker(**w) for w in workers]
        return []

    def get(self, worker_id: str) -> Worker:
        """获取 Worker 详情"""
        data = self.http.get(f"/api/{self.resource}/{worker_id}")
        return Worker(**data)

    def get_online(self) -> List[Worker]:
        """获取在线 Worker"""
        workers = self.list()
        return [w for w in workers if w.status == WorkerStatus.ONLINE]

    def get_available(self) -> List[Worker]:
        """获取可用 Worker（负载未满）"""
        workers = self.list()
        return [w for w in workers if w.status == WorkerStatus.ONLINE and w.current_load < w.max_load]

    def get_by_project(self, project_key: str) -> List[Worker]:
        """获取指定项目的 Worker"""
        workers = self.list()
        return [
            w
            for w in workers
            if w.status == WorkerStatus.ONLINE
            and (project_key in w.supported_projects or "*" in w.supported_projects)
        ]

    def get_stats(self) -> dict:
        """获取 Worker 状态统计"""
        workers = self.list()
        return {
            "total": len(workers),
            "online": len([w for w in workers if w.status == WorkerStatus.ONLINE]),
            "offline": len([w for w in workers if w.status == WorkerStatus.OFFLINE]),
            "busy": len([w for w in workers if w.status == WorkerStatus.BUSY]),
        }
